hibernate-core/src/main/java/org/hibernate/query/hql/internal/QualifiedJoinPathConsumer.java

```diff
@@ -207,6 +207,10 @@ else if ( fetch ) {
 				}
 			}
 		}
+		if ( !(subPathSource instanceof SqmJoinable) ) {
+			throw new SemanticException( "Joining on basic value map key elements is not supported!",
```

Comment:
The error is not correct. At this point, we don't know if the `subPathSource` is a key, we only know it is not joinable, from which we can at most assume that it must be basic. If you want to understand whether it is a map key, you must check `CollectionPart.Nature.INDEX.getName().equals( subPathSource.getPartName() )`